File: Python/1951-2000/1971. Find if Path Exists in Graph.py
# https://leetcode.com/problems/find-if-path-exists-in-graph/

class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph =[] 
        for i in range(0, n):
            graph.append([]) # Populate Graph

        # Create Graph
        for index in range(0, len(edges)):
            graph[edges[index][0]].append(edges[index][1])
            graph[edges[index][1]].append(edges[index][0])

        storage, visited = [source], {source}

        while(len(storage) > 0):

            current = storage.pop()
            if current == destination:
                return True # Destination found

            for nbrs in graph[current]:
                if nbrs not in visited:
                    visited.add(nbrs)
                    storage.append(nbrs) # add neighbours to queue and mark as visited

        return False # Destination not found

File: Javascript/1951-2000/1971. Find if Path Exists in Graph.js
// https://leetcode.com/problems/find-if-path-exists-in-graph/

var validPath = function(n, edges, source, destination) {
    const graph =[] 
    for(let i=0; i<n; i++) graph.push([]); // Populate Graph

    // Create Graph
    for(let index=0; index<edges.length; index++){
        graph[edges[index][0]].push(edges[index][1]);
        graph[edges[index][1]].push(edges[index][0]);
    }

    const storage = [source];
    const visited = new Set([source]); 

    while(storage.length > 0){

        let current =storage.pop(storage.length-1);
        if(current == destination) return true; // Destination found

        for(let nbrs of graph[current]){
            if(!visited.has(nbrs)){
                visited.add(nbrs);
                storage.push(nbrs); // add neighbours to queue and mark as visited
            }
        }
    }

    return false; // Destination not found
};
